### PR TITLE
Support non-blocking features using a shared codebase for sync and async clients

### PR DESCRIPTION
**Pull request by umarbinzahid**
This PR implements the `OpImpl` trait to support synchronous features in the async client while sharing a common codebase for both sync and async clients.

The following synchronous features are implemented with their unit test cases:
1. Set Socket Options.
2. Stop Provisioning Mode.
3. Enable Access Point.
4. Disable Access Point.
5. Get WINC Mac Address.
6. Get Firmware Version of WINC module.

All features are tested on hardware.

## Summary by Sourcery

Introduce a shared synchronous operation abstraction used by both sync and async clients to support non-blocking, single-shot operations such as firmware/MAC queries, provisioning, AP control, and socket configuration.

New Features:
- Expose synchronous-style helpers on the async client to set socket options, control provisioning mode, toggle access point mode, and query firmware and MAC address.
- Add a SyncOp wrapper to represent synchronous operations over the existing OpImpl state machine, enabling reuse between sync and async clients.

Enhancements:
- Refactor the sync client to route firmware version retrieval, provisioning stop, access point enable/disable, and socket option configuration through the shared SyncOp implementation and a common poll_once helper.
- Add generic poll_once helpers to both sync and async clients to execute a single OpImpl step and normalize ContinueOperation handling.

Tests:
- Extend async and sync client tests to cover the new SyncOp behaviors, including AP state transitions, socket option handling (UDP/TCP/SSL, timeouts, multicast), MAC address retrieval success/failure, and firmware version retrieval edge cases.

## Release Notes

* **New Features**
  * Added synchronous methods for configuring socket options, controlling access point provisioning mode, and retrieving device information (MAC address and firmware version).
  * All new operations return proper error handling without requiring async calls.

* **Refactor**
  * Internal operation execution refactored to improve synchronous operation handling and consistency across the client API.

### CODE REVIEW

File: winc-rs/src/net_ops/module.rs [hunk 1]
```diff
@@ -17,6 +27,76 @@ pub(crate) struct StationMode<'a> {
...
+                    }
+                    TcpSockOpts::ReceiveTimeout(timeout) => {
+                        // Receive timeout are handled by winc stack not by module.
+                        sock.set_recv_timeout(*timeout);
+                    }
+                }
+            }
+        }
+        Ok(Some(()))
+    }
```
[the remark this answers is not in the document]
Comment (umarbinzahid) [reply winc-rs/src/net_ops/module.rs]: Helper functions are added to handle the TCP and UDP socket options separately.

File: winc-rs/src/net_ops/module.rs [hunk 2]
```diff
@@ -1,14 +1,57 @@
...
+use crate::manager::{SslSockConfig, SslSockOpts};
+
+#[cfg(feature = "flash-rw")]
+use crate::manager::BootMode;
 
 // 1 minute max, if no other delays are added
 const AP_CONNECT_TIMEOUT_MILLISECONDS: u32 = 60_000;
 
+/// Non-blocking operations type.
+enum NoPollOpType<'a> {
```
Comment (kaidokert) [inline winc-rs/src/net_ops/module.rs]: Interesting approach and looks good to me.

The only quibble: those are essentially synchronous ops ( as opposed to asynchcronous ops that are most socket ops that need to wait for a response from the module ). Perhaps instead of "NoPoll" we can call them "SynchronousOpType" or "SyncOpType" ? Removes a negative and is perhaps easier to reason.

**Comment by kaidokert** [comment]
Overall looks like a good approach. Just the naming quibble: "NoPoll" vs "Sync" or "Synchronous" prefix for the struct names.

File: winc-rs/src/net_ops/module.rs (hunk 2 above)
Comment (umarbinzahid) [reply winc-rs/src/net_ops/module.rs]: I have updated the names to `SyncOpType` and `SyncOp`.

File: winc-rs/src/net_ops/module.rs [hunk 3]
```diff
@@ -17,6 +60,231 @@ pub(crate) struct StationMode<'a> {
...
+        Err(StackError::InvalidState)
+    }
+
+    /// Returns the WINC MAC address if it has been acquired.
+    ///
+    /// # Returns
+    ///
+    /// * `Ok(FirmwareInfo)` - WINC MAC address if available.
+    /// * `Err(StackError)` - If the MAC address request has not been made or if it failed.
+    pub(crate) fn retrieve_winc_mac_address(&mut self) -> Result<MacAddress, StackError> {
```
[the remark this answers is not in the document]
Comment (umarbinzahid) [reply winc-rs/src/net_ops/module.rs:268]: Doc is updated.

**Review by kaidokert — APPROVED** [review]
This seems ready to land
